### PR TITLE
UCT/IB: Add query_device compat layer

### PR DESCRIPTION
**Pull request by Artemy-Mellanox**

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10476/ for details (Mellanox internal link).

File: src/uct/ib/mlx5/ib_mlx5_md.c [hunk 1]
```diff
@@ -199,6 +199,57 @@ static ucs_mpool_ops_t uct_ib_mlx5_dbrec_ops = {
     .obj_cleanup   = NULL
 };
 
+static ucs_status_t uct_ib_mlx5dv_check_odp(uct_ib_mlx5_md_t *md, void *cap) {
```
Comment (yosefe) [inline src/uct/ib/mlx5/ib_mlx5_md.c]: { on next line

File: src/uct/ib/base/ib_verbs.h [hunk 2]
```diff
@@ -63,23 +57,73 @@
...
+    if (ret != 0) {
+        ucs_error("ibv_exp_query_device(%s) returned %d: %m",
+                  ibv_get_device_name(ctx->device), ret);
+        return UCS_ERR_IO_ERROR;
+    }
+
+    return UCS_OK;
+}
+#elif HAVE_DECL_IBV_QUERY_DEVICE_EX
+#  define IBV_DEV_ATTR(_dev, _attr)        ((_dev)->dev_attr.orig_attr._attr)
```
Comment (yosefe) [inline src/uct/ib/base/ib_verbs.h]: space line before

File: src/uct/ib/base/ib_verbs.h [hunk 3]
```diff
@@ -63,23 +57,73 @@
...
+    if (ret != 0) {
+        ucs_error("ibv_query_device_ex(%s) returned %d: %m",
+                  ibv_get_device_name(ctx->device), ret);
+        return UCS_ERR_IO_ERROR;
+    }
+
+    return UCS_OK;
+}
+
+#else
```
Comment (yosefe) [inline src/uct/ib/base/ib_verbs.h]: space line after

File: src/uct/ib/base/ib_verbs.h [hunk 4]
```diff
@@ -63,23 +57,73 @@
...
+
+    ret = ibv_query_device(ctx, attr);
+    if (ret != 0) {
+        ucs_error("ibv_query_device(%s) returned %d: %m",
+                  ibv_get_device_name(ctx->device), ret);
+        return UCS_ERR_IO_ERROR;
+    }
+
+    return UCS_OK;
+}
```
Comment (yosefe) [inline src/uct/ib/base/ib_verbs.h]: space line after

File: src/uct/ib/base/ib_verbs.h [hunk 5]
```diff
@@ -63,23 +57,73 @@
...
 #  define IBV_EXP_DEVICE_ATTR_SET_COMP_MASK(_attr)
-#  define IBV_EXP_PORT_ATTR_SET_COMP_MASK(_attr)
 
 #else
 #  define IBV_SHARED_MR_ACCESS_FLAGS(_shared_mr)    ((_shared_mr)->access)
+#endif /* HAVE_VERBS_EXP_H */
+
+/* Read device properties */
+#if HAVE_DECL_IBV_EXP_QUERY_DEVICE
+#  define IBV_DEV_ATTR(_dev, _attr)        ((_dev)->dev_attr._attr)
```
Comment (yosefe) [inline src/uct/ib/base/ib_verbs.h]: space line before

File: src/uct/ib/mlx5/ib_mlx5_md.c [hunk 6]
```diff
@@ -199,6 +199,57 @@ static ucs_mpool_ops_t uct_ib_mlx5_dbrec_ops = {
...
+    }
+
+    odp = UCT_IB_MLX5DV_ADDR_OF(query_hca_cap_out, out, capability);
+    UCT_IB_MLX5DV_SET(query_hca_cap_in, in, opcode, UCT_IB_MLX5_CMD_OP_QUERY_HCA_CAP);
+    UCT_IB_MLX5DV_SET(query_hca_cap_in, in, op_mod, UCT_IB_MLX5_HCA_CAP_OPMOD_GET_MAX |
+                                                   (UCT_IB_MLX5_CAP_ODP << 1));
+    ret = mlx5dv_devx_general_cmd(md->super.dev.ibv_context, in, sizeof(in),
+                                  out, sizeof(out));
+    if (ret != 0) {
+        return UCS_ERR_IO_ERROR;
```
Comment (yosefe) [inline src/uct/ib/mlx5/ib_mlx5_md.c]: ucs_error

File: src/uct/ib/mlx5/ib_mlx5_md.c [hunk 7]
```diff
@@ -199,6 +199,57 @@ static ucs_mpool_ops_t uct_ib_mlx5_dbrec_ops = {
...
+            md->super.config.odp.max_size = 1ul << 55;
+        } else {
+            md->super.config.odp.max_size = 1ul << 28;
+        }
+    }
+
+    if (UCT_IB_MLX5DV_GET(cmd_hca_cap, cap, fixed_buffer_size) &&
+        UCT_IB_MLX5DV_GET(cmd_hca_cap, cap, null_mkey) &&
+        UCT_IB_MLX5DV_GET(cmd_hca_cap, cap, umr_extended_translation_offset)) {
+        md->super.dev.flags |= UCT_IB_DEVICE_FLAG_ODP_IMPLICIT;
```
Comment (yosefe) [inline src/uct/ib/mlx5/ib_mlx5_md.c]: /* assume a devx cabable driver also implements ODP */
maybe also check that ODP for RC,UD is supported in ibv_device_attr_ex?

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10508/ for details (Mellanox internal link).

**Comment by Artemy-Mellanox** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10524/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10530/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10533/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10545/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
bot:pipe:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10549/ for details (Mellanox internal link).

File: src/uct/ib/mlx5/ib_mlx5_md.c [hunk 8]
```diff
@@ -199,38 +199,41 @@ static ucs_mpool_ops_t uct_ib_mlx5_dbrec_ops = {
...
 
     if (!UCT_IB_MLX5DV_GET(cmd_hca_cap, cap, pg)) {
-        return UCS_OK;
+        goto no_odp;
     }
 
     odp = UCT_IB_MLX5DV_ADDR_OF(query_hca_cap_out, out, capability);
     UCT_IB_MLX5DV_SET(query_hca_cap_in, in, opcode, UCT_IB_MLX5_CMD_OP_QUERY_HCA_CAP);
-    UCT_IB_MLX5DV_SET(query_hca_cap_in, in, op_mod, UCT_IB_MLX5_HCA_CAP_OPMOD_GET_MAX |
+    UCT_IB_MLX5DV_SET(query_hca_cap_in, in, op_mod, UCT_IB_MLX5_HCA_CAP_OPMOD_GET_CUR |
```
Comment (yosefe) [inline src/uct/ib/mlx5/ib_mlx5_md.c]: 👍

**Comment by yosefe** [comment]
bot:mlx:retest

**Comment by Artemy-Mellanox** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10603/ for details (Mellanox internal link).

**Comment by Artemy-Mellanox** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10610/ for details (Mellanox internal link).

**Comment by Artemy-Mellanox** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10615/ for details (Mellanox internal link).

**Pull request merged**